subnet/kube/kube.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2016 flannel authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"time"
+
+	"github.com/coreos/flannel/pkg/ip"
+	"github.com/coreos/flannel/subnet"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+var (
+	ErrUnimplemented = errors.New("unimplemented")
+	kubeSubnetCfg    *subnet.Config
+
+	defaultCniConf = []byte(`
+{
+    "name": "cbr0",
+    "type": "flannel",
+    "delegate": {
+      "isDefaultGateway": true
+    }
+}`)
+)
+
+const (
+	resyncPeriod = 5 * time.Minute
+
+	subnetKubeManagedAnnotation = "flannel.alpha.coreos.com/kube-subnet-manager"
+	backendDataAnnotation       = "flannel.alpha.coreos.com/backend-data"
+	backendTypeAnnotation       = "flannel.alpha.coreos.com/backend-type"
+	backendPublicIPAnnotation   = "flannel.alpha.coreos.com/public-ip"
+
+	cniConfPath       = "/etc/kube-flannel/cni-conf.json"
+	cniConfTargetPath = "/etc/cni/net.d/10-flannel.conf"
+	netConfPath       = "/etc/kube-flannel/net-conf.json"
+)
+
+type kubeSubnetManager struct {
+	client         clientset.Interface
+	nodeName       string
+	nodeStore      cache.StoreToNodeLister
+	nodeController *framework.Controller
+	subnetConf     *subnet.Config
+}
+
+func NewSubnetManager() (subnet.Manager, error) {
+	cfg, err := restclient.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize inclusterconfig: %v", err)
+	}
+	c, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize client: %v", err)
+	}
+
+	podName := os.Getenv("POD_NAME")
+	podNamespace := os.Getenv("POD_NAMESPACE")
+	if podName == "" || podNamespace == "" {
+		return nil, fmt.Errorf("env variables POD_NAME and POD_NAMESPACE must be set")
+	}
+
+	pod, err := c.Pods(podNamespace).Get(podName)
+	nodeName := pod.Spec.NodeName
```

Comment:
@mikedanese @aaronlevy 
Since we have the node name here, could we as a intermediate step while waiting for componentconfig parse the two first bytes in ``.spec.PodCIDR` and make a `/16` network of it?

While I think the configmap is good, I also think that the one-line `kubectl apply -f https://k8s.io/flannel-network.yaml` is more important, and should
a) default to something (like weave does [here](https://github.com/weaveworks/weave-kube/blob/master/weave-daemonset.yaml#L22)) along with the `vxlan` backend.
b) guess it and hope for the best
c) entirely skip the configmap and use annotations on the Pod instead, since it is only read once (updating the configmap doesn't update the flannel conf anyway)

WDYT?